server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/FilterTest.java

```diff
@@ -240,6 +240,134 @@ public void setFilterShouldOverwritePreviouslyStoredRules() {
+                "      \"name\": \"some name\"," +
+                "      \"condition\": {" +
+                "        \"field\": \"subject\"," +
+                "        \"comparator\": \"contains\"," +
+                "        \"value\": \"question\"" +
+                "      }," +
+                "      \"action\": {" +
+                "        \"appendIn\": {" +
+                "          \"mailboxIds\": [\"" + mailbox.serialize() + "\"]" +
+                "        }" +
+                "      }" +
+                "    }" +
+                "  ]}, " +
+                "\"#0\"" +
+                "]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .body(ARGUMENTS + ".type", equalTo("invalidArguments"))
+            .body(ARGUMENTS + ".description", equalTo("`id` is mandatory"));
+    }
+
+    @Test
+    public void setFilterShouldRejectRuleWithInvalidRuleCondition() {
+        MailboxId mailbox = randomMailboxId();
+
+        given()
+            .header("Authorization", accessToken.asString())
+            .body("[[" +
+                "  \"setFilter\", " +
+                "  {" +
+                "    \"singleton\": [" +
+                "    {" +
+                "      \"id\": \"3000-34e\"," +
+                "      \"name\": \"some name\"," +
+                "      \"condition\": null," +
+                "      \"action\": {" +
+                "        \"appendIn\": {" +
+                "          \"mailboxIds\": [\"" + mailbox.serialize() + "\"]" +
+                "        }" +
+                "      }" +
+                "    }" +
+                "  ]}, " +
+                "\"#0\"" +
+                "]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .body(ARGUMENTS + ".type", equalTo("anErrorOccurred"))
+            .body(ARGUMENTS + ".description", equalTo("Failed to retrieve filter"));
+    }
+
+    @Test
+    public void setFilterShouldRejectRuleWithInvalidRuleAction() {
+        MailboxId mailbox = randomMailboxId();
+
+        given()
+            .header("Authorization", accessToken.asString())
+            .body("[[" +
+                "  \"setFilter\", " +
+                "  {" +
+                "    \"singleton\": [" +
+                "    {" +
+                "      \"id\": \"3000-34e\"," +
+                "      \"name\": \"some name\"," +
+                "      \"condition\": {" +
+                "        \"field\": \"subject\"," +
+                "        \"comparator\": \"contains\"," +
+                "        \"value\": \"question\"" +
+                "      }," +
+                "      \"action\": null" +
+                "    }" +
+                "  ]}, " +
+                "\"#0\"" +
+                "]]")
+        .when()
+            .post("/jmap")
+        .then()
+             .body(ARGUMENTS + ".type", equalTo("anErrorOccurred"))
+             .body(ARGUMENTS + ".description", equalTo("Failed to retrieve filter"));
```

Comment:
I would be expecting an invalidArguments here too.... The error is misleading and do not even mention the fact that "action" should not be null.

Reply:
Yes, just waiting for others opinion.